snap/gadget.go

```diff
@@ -249,3 +496,105 @@ func (e *editionNumber) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	*e = editionNumber(u)
 	return nil
 }
+
+// GadgetSize describes the size of a structure item or an offset within the
+// structure.
+type GadgetSize uint64
+
+const (
+	SizeMiB = GadgetSize(2 << 20)
+	SizeGiB = GadgetSize(2 << 30)
+)
+
+func (s *GadgetSize) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var gs string
+	if err := unmarshal(&gs); err != nil {
+		return errors.New(`failed to unmarshal`)
+	}
+
+	var err error
+	*s, err = ParseGadgetSize(gs)
+	if err != nil {
+		return fmt.Errorf("cannot parse size %q: %v", gs, err)
+	}
+	return err
+}
+
+// ParseGadgetSize parses a string expressing size in gadget declaration. The
+// accepted format is one of: <bytes> | <bytes/2^20>M | <bytes/2^30>G.
+func ParseGadgetSize(gs string) (GadgetSize, error) {
+	number, unit, err := strutil.SplitUnit(gs)
+	if err != nil {
+		return 0, err
+	}
+	if number < 0 {
+		return 0, errors.New("size cannot be negative")
+	}
+	var size GadgetSize
+	switch unit {
+	case "M":
+		// MiB
+		size = GadgetSize(number) * SizeMiB
+	case "G":
+		// GiB
+		size = GadgetSize(number) * SizeGiB
+	case "":
+		// straight bytes
+		size = GadgetSize(number)
+	default:
+		return 0, fmt.Errorf("invalid suffix %q", unit)
+	}
+	return size, nil
+}
+
+// GadgetRelativeOffset describes an offset where structure data is written at.
+// The position can be specified as byte-offset relative to the start of another
+// named structure.
+type GadgetRelativeOffset struct {
+	RelativeTo string
+	Offset     GadgetSize
+}
+
+// ParseGadgetRelativeOffset parses a string describing an offset that can be
+// expressed relative to a named structure, with the format: [<name>+]<size>.
+func ParseGadgetRelativeOffset(grs string) (*GadgetRelativeOffset, error) {
+	toWhat := ""
+	sizeSpec := grs
+	if idx := strings.IndexRune(grs, '+'); idx != -1 {
+		toWhat, sizeSpec = grs[:idx], grs[idx+1:]
+		if toWhat == "" {
+			return nil, errors.New("missing volume name")
+		}
+	}
+	if sizeSpec == "" {
+		return nil, errors.New("missing offset")
+	}
+
+	size, err := ParseGadgetSize(sizeSpec)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse offset %q: %v", sizeSpec, err)
+	}
+	if size > 4*SizeGiB {
```

Comment:
The comment below is redundant. Can you please add a reference to the origin of the limit?